cpp/src/strings/utilities.cuh

```diff
@@ -102,12 +105,103 @@ auto make_strings_children(
   auto const bytes = cudf::detail::get_value<int32_t>(offsets_view, strings_count, stream);
   std::unique_ptr<column> chars_column =
     create_chars_child_column(strings_count, bytes, stream, mr);
-  size_and_exec_fn.d_chars = chars_column->mutable_view().template data<char>();
-  for_each_fn(size_and_exec_fn);
+
+  // Execute the function fn again to fill the chars column.
+  // Note that if the output chars column has zero size, the function fn should not be called to
+  // avoid accidentally overwriting the offsets.
+  if (bytes > 0) {
+    size_and_exec_fn.d_chars = chars_column->mutable_view().template data<char>();
+    for_each_fn(size_and_exec_fn);
+  }
 
   return std::make_pair(std::move(offsets_column), std::move(chars_column));
 }
 
+/**
+ * @brief Creates child offsets, chars columns and null mask, null count of a strings column by
+ * applying the template function that can be used for computing the output size of each string as
+ * well as create the output.
+ *
+ * @tparam SizeAndExecuteFunction Function must accept an index and return a size.
+ *         It must have members `d_offsets`, `d_chars`, and `d_validities` which are set to memory
+ *         containing the offsets column, chars column and string validities during write.
+ *
+ * @param size_and_exec_fn This is called twice. Once for the output size of each string, which is
+ *                         written into the `d_offsets` array. After that, `d_chars` is set and this
+ *                         is called again to fill in the chars memory. The `d_validities` array may
+ *                         be modified to set the value `0` for the corresponding rows that contain
+ *                         null string elements.
+ * @param strings_count Number of strings.
+ * @param mr Device memory resource used to allocate the returned columns' device memory.
+ * @param stream CUDA stream used for device memory operations and kernel launches.
+ * @return offsets child column, chars child column, null_mask, and null_count for a strings column.
+ */
+template <typename SizeAndExecuteFunction>
+std::tuple<std::unique_ptr<column>, std::unique_ptr<column>, rmm::device_buffer, size_type>
+make_strings_children_with_null_mask(
+  SizeAndExecuteFunction size_and_exec_fn,
+  size_type strings_count,
+  rmm::cuda_stream_view stream        = rmm::cuda_stream_default,
+  rmm::mr::device_memory_resource* mr = rmm::mr::get_current_device_resource())
+{
```

Comment:
Considering I foresee this is only to be used by the `concatenate_list_elements` I wonder if it should not be added to this include file since this file is included by many, many other source files in libcudf increasing compile time.

Reply:
I thought about that too, but I was not sure whether we will need this utility function in the future for other APIs. Currently, I'm working on row-wise lists concatenation and lists column interleaving, which also include operations for string column. They are about to be completed soon and I'll let you know if this utility is needed there.

Reply:
@davidwendt I think we should leave the utility function in that header, as it would be many potential usages. In addition, I would like to rewrite the implementation in `combine.cu` (in a separate PR) to use it---that would make the code a lot shorter.

Reply:
I'm ok with leaving it as utility if you think it will be valuable to your other work. 
If that work is outside of `cpp/src/strings/` then we may want to move this utility to `cpp/include/cudf/strings/detail/utilities.cuh` instead.
I'm not ok with rewriting combine/concatenate.cu to use it.